### ISSUE TITLE
DCP probe is asynchronous and can race login manager

### ISSUE THREAD

**Issue by levihuayuzhang** [AsahiLinux/linux#87]
Ok, I find the solution (just add DRM support in initramfs):
1. Edit the `mkinitcpio` HOOK file `asahi` using `vim /usr/lib/initcpio/install/asahi`:
2. add following content:
```sh
# For DRM
    map add_module drm_dma_helper? asahi? apple_piodma? appledrm? \
	 apple_dcp?  panel-simple? virtio-gpu? gpu-sched?
```
3. generate a new initramfs image for `asahi-linux-edge`:
```sh
mkinitcpio -p linux-asahi-edge
```
4. The Gnome use Wayland by default, but if you want to ensure, edit `vim /etc/gdm/custom.conf`:
```sh
[daemon]
WaylandEnable=true
DefaultSession=gnome-wayland.desktop
```
5. Fresh reboot, then then GPU accelerated Wayland Gnome would be works (without log out). 
6. (No need to modify grub settings or use kernel parameter)

I'm wondering if the DRM features become stable enough (in the future or maybe now?), maybe we can add the DRM modules to initramfs to improve `default` experience (Cause the Wayland seems works better than X11 with Asahi for now)?

_Originally posted by @levihuayuzhang in https://github.com/AsahiLinux/linux/issues/83#issuecomment-1345614438_

**Comment by marcan** [comment]
This is not a valid solution. The GPU modules are not supposed to be in the initramfs. If you are having problems with the modules in the rootfs, that means something you are running has a race condition and is trying to load before devices properly probe. That means that needs to be fixed, not by adding the modules to the initramfs.

**Comment by levihuayuzhang** [comment]
> This is not a valid solution. The GPU modules are not supposed to be in the initramfs. If you are having problems with the modules in the rootfs, that means something you are running has a race condition and is trying to load before devices properly probe. That means that needs to be fixed, not by adding the modules to the initramfs.

Yes, totally agree. Above is a early KMS start I found in Arch Linux WIKI. 
I'm just wondering our DRM module is functioning (or loading) "right" for Wayland session to start: 
Our KMS inplementation should already set a right mode of gpu for basic (like native resolution in the framebuffer). But maybe other additional newer tech requied (Required by Wayland session) should be enabled? (I just know the DRM is requied by Wayland, and DRM early loading did fix the wayland-session issue)<https://wayland-book.com/introduction/high-level-design.html>
Or systemd start GDM before modules are well prepared?

I'll try to dive into it.

**Comment by marcan** [comment]
I think right now DCP probes asynchronously, which can introduce a race with GDM. I've brought it up with jannau, let's see if we can make it synchronous and whether that fixes your problem.

**Comment by jannau** [comment]
This should be fixed by the component based probing of the display driver merged in `asahi-v6.2-rc3-6`.
Let's keep this open until it lands in `asahi` and is confirmed as fixed.

**Comment by marcan** [comment]
This should be fixed now.
